File: website/api/support/Polyline.php
<?php

/**
 * Google encoded-polyline codec, precision 6.
 *
 * Precision 5 (the classic Maps default) quantises to ~1.1 m, which is coarse
 * enough to visibly stair-step a snapped road at street zoom. Precision 6 is
 * ~11 cm and is what both OSRM (`geometries=polyline6`) and Valhalla emit, so
 * six is not a choice here so much as the wire format the matchers already
 * speak.
 *
 * Why encode at all: a matched day is a few thousand vertices. As JSON floats
 * that is ~40 KB; as polyline6 it is ~8 KB, and the client decodes it in one
 * pass instead of allocating a list of two-element lists.
 */
final class Polyline
{
    private const PRECISION = 6;

    /**
     * @param array<int, array{0: float, 1: float}> $points [lat, lng] pairs;
     *        extra tuple members are ignored, so a route tuple can be passed in
     *        unchanged.
     */
    public static function encode(array $points): string
    {
        $factor = 10 ** self::PRECISION;
        $out = '';
        $prevLat = 0;
        $prevLng = 0;

        foreach ($points as $point) {
            $lat = (int) round(((float) $point[0]) * $factor);
            $lng = (int) round(((float) $point[1]) * $factor);

            $out .= self::chunk($lat - $prevLat) . self::chunk($lng - $prevLng);

            // Deltas are taken against the *rounded* previous value, never the
            // original float, or the error compounds along the line.
            $prevLat = $lat;
            $prevLng = $lng;
        }

        return $out;
    }

    /** @return array<int, array{0: float, 1: float}> */
    public static function decode(string $encoded): array
    {
        $factor = 10 ** self::PRECISION;
        $length = strlen($encoded);
        $index = 0;
        $lat = 0;
        $lng = 0;
        $out = [];

        while ($index < $length) {
            $lat += self::unchunk($encoded, $index, $length);
            $lng += self::unchunk($encoded, $index, $length);
            $out[] = [$lat / $factor, $lng / $factor];
        }

        return $out;
    }

    /** Zig-zag the sign into bit 0, then emit 5 bits per character. */
    private static function chunk(int $value): string
    {
        $v = $value < 0 ? ~($value << 1) : ($value << 1);
        $out = '';

        while ($v >= 0x20) {
            $out .= chr((0x20 | ($v & 0x1f)) + 63);
            $v >>= 5;
        }

        return $out . chr($v + 63);
    }

    private static function unchunk(string $encoded, int &$index, int $length): int
    {
        $result = 0;
        $shift = 0;
        $byte = 0;

        do {
            if ($index >= $length) {
                break;
            }
            $byte = ord($encoded[$index++]) - 63;
            $result |= ($byte & 0x1f) << $shift;
            $shift += 5;
        } while ($byte >= 0x20);

        return ($result & 1) ? ~($result >> 1) : ($result >> 1);
    }
}

File: mobile_app/lib/core/utils/polyline.dart
/// Google encoded-polyline codec, precision 6.
///
/// The server sends road-matched geometry this way rather than as JSON
/// coordinate pairs: a matched day is a few thousand vertices, which is ~40 KB
/// of floats and ~8 KB encoded, and it decodes in one pass instead of
/// allocating a list of two-element lists.
///
/// Precision 6 (~11 cm), not the classic 5 (~1.1 m) — five visibly stair-steps
/// a snapped road at street zoom, and six is what OSRM's `polyline6` and
/// Valhalla both emit.
class PolylineCodec {
  const PolylineCodec._();

  static const double _factor = 1e6;

  /// Returns `[latitude, longitude]` pairs.
  ///
  /// A truncated or malformed string yields the vertices decoded so far rather
  /// than throwing: geometry is a rendering input, and half a route drawn beats
  /// a crashed map screen.
  static List<List<double>> decode(String encoded) {
    final List<List<double>> out = <List<double>>[];
    int index = 0;
    int lat = 0;
    int lng = 0;

    while (index < encoded.length) {
      final (int dLat, int afterLat) = _value(encoded, index);
      if (afterLat == index) break;

      final (int dLng, int afterLng) = _value(encoded, afterLat);

      lat += dLat;
      lng += dLng;
      index = afterLng;

      out.add(<double>[lat / _factor, lng / _factor]);
    }

    return out;
  }

  /// One varint, and where it ended.
  static (int, int) _value(String encoded, int index) {
    int result = 0;
    int shift = 0;
    int byte = 0;

    while (index < encoded.length) {
      byte = encoded.codeUnitAt(index++) - 63;
      result |= (byte & 0x1f) << shift;
      shift += 5;
      if (byte < 0x20) break;
    }

    // Bit 0 carries the sign (zig-zag), so an odd value is a negative delta.
    return ((result & 1) != 0 ? ~(result >> 1) : (result >> 1), index);
  }
}
